MH-13133, TypeError: Cannot read property 'results' of null

lkiesow opened the pull request:
Especially on slow networks, the admin interface throws exceptions since
it does not get all necessary resources but tries to access them anyway.

This patch checks if the received data is valid. It also prevents
Opencast from hiding errors in case there really are problems.

lkiesow committed:
MH-13133, TypeError: Cannot read property 'results' of null

Especially on slow networks, the admin interface throws exceptions since
it does not get all necessary resources but tries to access them anyway.

This patch checks if the received data is valid. It also prevents
Opencast from hiding errors in case there really are problems.

lkiesow commented:
Btb. the error was easy to reproduce when throttling the network speed in Chrome (Developer Tools → Network →Throttling → Slow 3G) but you also get it from time to time on a fast network.

staubesv commented on modules/admin-ui/src/main/webapp/scripts/shared/services/resourceHelperService.js:
```diff
@@ -44,28 +44,25 @@ angular.module('adminNg.services')
 
             // guard against empty dataset
-            if (angular.isUndefined(data.results)) {
+            if (data === null || angular.isUndefined(data.results)) {
```
@lkiesow As this behaviour seems weird to me, I digged somewhat into what is causing JSON.parse to result in _null_. The cause is that responseBody is already _null_ and _JSON.parse(null)_ results in _null_.
This happens, for example, if a request is cancelled (or when ever the server responds with an empty body).

Maybe it would be better to check _(responseBody === null_) at the beginning of the method to make the code easier to understand.

What do you think?

lkiesow replied:
That would only catch half of the cases: If the returned JSON would be null (Important: JSON=null not JS-Value=null), you would still run into the same error. I can add a second check before parsing the response if you want me to. Though that's then more or less redundant.

Though in any case, `angular.isObject` may be a better chice than a check for `null`. I'll update this pull request.

Please let me know if you want me to also include the additional `null` check before parsing the JSON.

lkiesow replied:
Thinking about this again, `angular.isObject` might hide some errors and it probably makes more sense to keep the `null` check. Still, let me know if you want another one before the `JSON.parse`.

lkiesow committed:
MH-13133, Simplify resourceHelperService

This patch strives to make the admin interface resourceHelperService
more readable and easier to understand.

lkiesow commented:
I've tried to simplify this file a bit more. I've also included the additional `null` check.

staubesv commented:
Hmm... the whole code is actually quite messy (talking about the original code, not your changes): If the method would consider the HTTP status code and the backend would always deliver the data correctly formatted, there would be no need for any of those checks like "responseBody = null , data == null, data.results undefined, data.results is array or not)...

Anyway, this PR certainly does improve the code and fixes the undesirable console output.

lkiesow commented:
I agree. The backend not returning JSON in an error case while claiming it's JSON is already problematic. But unfortunately, that's a problem with nearly all admin-ng endpoints :-(

staubesv commented:
Well, the ResourceHelper is not supposed to parse anything if the backend does not return HTTP 200 as it expects a specific data format that only makes sense if the request succeeds...

Anyway, let's agree that this PR improves the situation and fixing "the Admin UI" might be out of scope ;-)
